From d60582dff925b9b749ab9cd4a07bf73e9f10633a Mon Sep 17 00:00:00 2001
From: Pulsar <pulsar@localhost>
Date: Thu, 13 Aug 2026 22:42:19 +0000
Subject: [PATCH] alioth: port Redline device fixes + USB fast charge

- OIS: fix tele5x module (check ois_name for sem1215) to prevent cam crash
- panel: K11A reset sequence fix (reset-powerdown-delay)
- power: smb1398 + bq2597x_dagu uintptr_t cast fixes
- backlight: remove ktz8866/dualktz8866 log spam
- misc: add force fast charge driver (sysfs /sys/kernel/fast_charge/force_fast_charge)
  hooks in smb5/smb-lib set_sdp_current (500mA -> 900mA when enabled)
- defconfig: enable CONFIG_FORCE_FAST_CHARGE
---
 .../qcom/dsi-panel-k11a-38-08-0a-dsc-cmd.dtsi |   3 +-
 arch/arm64/configs/vendor/alioth_defconfig    |   1 +
 drivers/misc/Kconfig                          |   6 +
 drivers/misc/Makefile                         |   1 +
 drivers/misc/fastchg.c                        | 103 ++++++++++++++++++
 drivers/power/supply/qcom/smb-lib.c           |  11 ++
 drivers/power/supply/qcom/smb5-lib-dagu.c     |  11 ++
 drivers/power/supply/qcom/smb5-lib-munch.c    |  11 ++
 drivers/power/supply/qcom/smb5-lib-psyche.c   |  11 ++
 drivers/power/supply/qcom/smb5-lib.c          |  10 ++
 .../power/supply/qcom_cas/smb1398-charger.c   |   2 +-
 drivers/power/supply/qcom_cas/smb5-lib.c      |  10 ++
 .../power/supply/ti/bq2597x_charger_dagu.c    |   4 +-
 drivers/video/backlight/dualktz8866.c         |   3 -
 drivers/video/backlight/ktz8866.c             |   3 -
 include/linux/fastchg.h                       |  22 ++++
 .../cam_sensor_module/cam_ois/cam_ois_core.c  |   7 +-
 17 files changed, 207 insertions(+), 12 deletions(-)
 create mode 100644 drivers/misc/fastchg.c
 create mode 100644 include/linux/fastchg.h

diff --git a/arch/arm64/boot/dts/vendor/qcom/dsi-panel-k11a-38-08-0a-dsc-cmd.dtsi b/arch/arm64/boot/dts/vendor/qcom/dsi-panel-k11a-38-08-0a-dsc-cmd.dtsi
index 53f745759..b43e7e4cc 100644
--- a/arch/arm64/boot/dts/vendor/qcom/dsi-panel-k11a-38-08-0a-dsc-cmd.dtsi
+++ b/arch/arm64/boot/dts/vendor/qcom/dsi-panel-k11a-38-08-0a-dsc-cmd.dtsi
@@ -23,7 +23,8 @@
 		qcom,mdss-dsi-lane-3-state;
 		qcom,mdss-dsi-dma-trigger = "trigger_sw";
 		qcom,mdss-dsi-mdp-trigger = "none";
-		qcom,mdss-dsi-reset-sequence = <0 10>, <1 10>;
+		qcom,mdss-dsi-reset-sequence = <0 12>, <1 10>;
+		qcom,mdss-dsi-reset-powerdown-delay = <10>;
 		qcom,mdss-pan-physical-width-dimension = <70>;
 		qcom,mdss-pan-physical-height-dimension = <155>;
 		qcom,mdss-dsi-te-pin-select = <1>;
diff --git a/arch/arm64/configs/vendor/alioth_defconfig b/arch/arm64/configs/vendor/alioth_defconfig
index 7484516a3..9af2d554b 100644
--- a/arch/arm64/configs/vendor/alioth_defconfig
+++ b/arch/arm64/configs/vendor/alioth_defconfig
@@ -1631,6 +1631,7 @@ CONFIG_QSEECOM=y
 # CONFIG_PCI_ENDPOINT_TEST is not set
 CONFIG_UID_SYS_STATS=y
 # CONFIG_QPNP_MISC is not set
+CONFIG_FORCE_FAST_CHARGE=y
 # CONFIG_HALL_AKM09970 is not set
 # CONFIG_MOTOR_DRV8846 is not set
 # CONFIG_QTI_XR_SMRTVWR_MISC is not set
diff --git a/drivers/misc/Kconfig b/drivers/misc/Kconfig
index e65d61884..eb747f686 100644
--- a/drivers/misc/Kconfig
+++ b/drivers/misc/Kconfig
@@ -11,6 +11,12 @@ config SPI_XIAOMI_TP
         help
         Say Y if want use xiaomi touch panel.
 
+config FORCE_FAST_CHARGE
+	bool "Force faster charge rate for USB"
+	default y
+	help
+	  This allows users to override default charge rate for USB
+
 config SENSORS_LIS3LV02D
 	tristate
 	depends on INPUT
diff --git a/drivers/misc/Makefile b/drivers/misc/Makefile
index 38d82779b..9167a3a6e 100644
--- a/drivers/misc/Makefile
+++ b/drivers/misc/Makefile
@@ -64,6 +64,7 @@ obj-$(CONFIG_MISC_RTSX)		+= cardreader/
 obj-$(CONFIG_UID_SYS_STATS)	+= uid_sys_stats.o
 obj-$(CONFIG_MEMORY_STATE_TIME)	+= memory_state_time.o
 obj-$(CONFIG_QPNP_MISC)		+= qpnp-misc.o
+obj-$(CONFIG_FORCE_FAST_CHARGE) += fastchg.o
 
 obj-$(CONFIG_OKL4_USER_VIRQ)	+= okl4-virq.o
 obj-$(CONFIG_OKL4_RINGBUF)	+= okl4-ringbuf.o
diff --git a/drivers/misc/fastchg.c b/drivers/misc/fastchg.c
new file mode 100644
index 000000000..4d4bb727f
--- /dev/null
+++ b/drivers/misc/fastchg.c
@@ -0,0 +1,103 @@
+/*
+ * Author: Chad Froebel <chadfroebel@gmail.com>
+ *
+ * Port to cheeseburger: engstk <eng.stk@sapo.pt>
+ *
+ * This software is licensed under the terms of the GNU General Public
+ * License version 2, as published by the Free Software Foundation, and
+ * may be copied, distributed, and modified under those terms.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ */
+
+/*
+ * Possible values for "force_fast_charge" are :
+ *
+ *   0 - Disabled (default)
+ *   1 - Force faster charge
+*/
+
+#include <linux/kobject.h>
+#include <linux/sysfs.h>
+#include <linux/fastchg.h>
+#include <linux/string.h>
+#include <linux/module.h>
+
+int force_fast_charge = 0;
+
+static int __init get_fastcharge_opt(char *ffc)
+{
+	if (strcmp(ffc, "0") == 0) {
+		force_fast_charge = 0;
+	} else if (strcmp(ffc, "1") == 0) {
+		force_fast_charge = 1;
+	} else {
+		force_fast_charge = 0;
+	}
+	return 1;
+}
+
+__setup("ffc=", get_fastcharge_opt);
+
+static ssize_t force_fast_charge_show(struct kobject *kobj, struct kobj_attribute *attr, char *buf)
+{
+	size_t count = 0;
+	count += sprintf(buf, "%d\n", force_fast_charge);
+	return count;
+}
+
+static ssize_t force_fast_charge_store(struct kobject *kobj, struct kobj_attribute *attr, const char *buf, size_t count)
+{
+	sscanf(buf, "%d ", &force_fast_charge);
+	if (force_fast_charge < 0 || force_fast_charge > 1)
+		force_fast_charge = 0;
+
+	return count;
+}
+
+static struct kobj_attribute force_fast_charge_attribute =
+__ATTR(force_fast_charge, 0664, force_fast_charge_show, force_fast_charge_store);
+
+static struct attribute *force_fast_charge_attrs[] = {
+&force_fast_charge_attribute.attr,
+NULL,
+};
+
+static struct attribute_group force_fast_charge_attr_group = {
+.attrs = force_fast_charge_attrs,
+};
+
+/* Initialize fast charge sysfs folder */
+static struct kobject *force_fast_charge_kobj;
+
+int force_fast_charge_init(void)
+{
+	int force_fast_charge_retval;
+
+	force_fast_charge_kobj = kobject_create_and_add("fast_charge", kernel_kobj);
+	if (!force_fast_charge_kobj) {
+			return -ENOMEM;
+	}
+
+	force_fast_charge_retval = sysfs_create_group(force_fast_charge_kobj, &force_fast_charge_attr_group);
+
+	if (force_fast_charge_retval)
+		kobject_put(force_fast_charge_kobj);
+
+	if (force_fast_charge_retval)
+		kobject_put(force_fast_charge_kobj);
+
+	return (force_fast_charge_retval);
+}
+
+void force_fast_charge_exit(void)
+{
+	kobject_put(force_fast_charge_kobj);
+}
+
+module_init(force_fast_charge_init);
+module_exit(force_fast_charge_exit);
diff --git a/drivers/power/supply/qcom/smb-lib.c b/drivers/power/supply/qcom/smb-lib.c
index b7548c38d..f85866fec 100644
--- a/drivers/power/supply/qcom/smb-lib.c
+++ b/drivers/power/supply/qcom/smb-lib.c
@@ -17,6 +17,10 @@
 #include "step-chg-jeita.h"
 #include "storm-watch.h"
 
+#ifdef CONFIG_FORCE_FAST_CHARGE
+#include <linux/fastchg.h>
+#endif
+
 #define smblib_err(chg, fmt, ...)		\
 	pr_err("%s: %s: " fmt, chg->name,	\
 		__func__, ##__VA_ARGS__)	\
@@ -869,6 +873,13 @@ static int set_sdp_current(struct smb_charger *chg, int icl_ua)
 	u8 icl_options;
 	const struct apsd_result *apsd_result = smblib_get_apsd_result(chg);
 
+#ifdef CONFIG_FORCE_FAST_CHARGE
+	if (force_fast_charge > 0 && icl_ua == USBIN_500MA)
+	{
+		icl_ua = USBIN_900MA;
+	}
+#endif
+
 	/* power source is SDP */
 	switch (icl_ua) {
 	case USBIN_100MA:
diff --git a/drivers/power/supply/qcom/smb5-lib-dagu.c b/drivers/power/supply/qcom/smb5-lib-dagu.c
index dbc29f1f5..f1ad9d797 100644
--- a/drivers/power/supply/qcom/smb5-lib-dagu.c
+++ b/drivers/power/supply/qcom/smb5-lib-dagu.c
@@ -24,6 +24,10 @@
 #include "dual_fuel_gauge_class.h"
 #endif
 
+#ifdef CONFIG_FORCE_FAST_CHARGE
+#include <linux/fastchg.h>
+#endif
+
 #define smblib_err(chg, fmt, ...)		\
 	pr_err("%s: %s: " fmt, chg->name,	\
 		__func__, ##__VA_ARGS__)	\
@@ -1840,6 +1844,13 @@ static int set_sdp_current(struct smb_charger *chg, int icl_ua)
 	u8 icl_options;
 	const struct apsd_result *apsd_result = smblib_get_apsd_result(chg);
 
+#ifdef CONFIG_FORCE_FAST_CHARGE
+	if (force_fast_charge > 0 && icl_ua == USBIN_500MA)
+	{
+		icl_ua = USBIN_900MA;
+	}
+#endif
+
 	/* power source is SDP */
 	switch (icl_ua) {
 	case USBIN_100MA:
diff --git a/drivers/power/supply/qcom/smb5-lib-munch.c b/drivers/power/supply/qcom/smb5-lib-munch.c
index ba8df9ca0..fad40580a 100755
--- a/drivers/power/supply/qcom/smb5-lib-munch.c
+++ b/drivers/power/supply/qcom/smb5-lib-munch.c
@@ -24,6 +24,10 @@
 #include "dual_fuel_gauge_class.h"
 #endif
 
+#ifdef CONFIG_FORCE_FAST_CHARGE
+#include <linux/fastchg.h>
+#endif
+
 #define smblib_err(chg, fmt, ...)		\
 	pr_err("%s: %s: " fmt, chg->name,	\
 		__func__, ##__VA_ARGS__)	\
@@ -1818,6 +1822,13 @@ static int set_sdp_current(struct smb_charger *chg, int icl_ua)
 	u8 icl_options;
 	const struct apsd_result *apsd_result = smblib_get_apsd_result(chg);
 
+#ifdef CONFIG_FORCE_FAST_CHARGE
+	if (force_fast_charge > 0 && icl_ua == USBIN_500MA)
+	{
+		icl_ua = USBIN_900MA;
+	}
+#endif
+
 	/* power source is SDP */
 	switch (icl_ua) {
 	case USBIN_100MA:
diff --git a/drivers/power/supply/qcom/smb5-lib-psyche.c b/drivers/power/supply/qcom/smb5-lib-psyche.c
index 92e57bcfc..39abedd41 100644
--- a/drivers/power/supply/qcom/smb5-lib-psyche.c
+++ b/drivers/power/supply/qcom/smb5-lib-psyche.c
@@ -24,6 +24,10 @@
 #include "dual_fuel_gauge_class.h"
 #endif
 
+#ifdef CONFIG_FORCE_FAST_CHARGE
+#include <linux/fastchg.h>
+#endif
+
 #define smblib_err(chg, fmt, ...)		\
 	pr_err("%s: %s: " fmt, chg->name,	\
 		__func__, ##__VA_ARGS__)	\
@@ -1821,6 +1825,13 @@ static int set_sdp_current(struct smb_charger *chg, int icl_ua)
 	u8 icl_options;
 	const struct apsd_result *apsd_result = smblib_get_apsd_result(chg);
 
+#ifdef CONFIG_FORCE_FAST_CHARGE
+	if (force_fast_charge > 0 && icl_ua == USBIN_500MA)
+	{
+		icl_ua = USBIN_900MA;
+	}
+#endif
+
 	/* power source is SDP */
 	switch (icl_ua) {
 	case USBIN_100MA:
diff --git a/drivers/power/supply/qcom/smb5-lib.c b/drivers/power/supply/qcom/smb5-lib.c
index 4ef0de56f..85f13edc5 100644
--- a/drivers/power/supply/qcom/smb5-lib.c
+++ b/drivers/power/supply/qcom/smb5-lib.c
@@ -23,6 +23,9 @@
 #ifdef CONFIG_DUAL_FUEL_GAUGE_BQ27Z561
 #include "dual_fuel_gauge_class.h"
 #endif
+#ifdef CONFIG_FORCE_FAST_CHARGE
+#include <linux/fastchg.h>
+#endif
 
 #define smblib_err(chg, fmt, ...)		\
 	pr_err("%s: %s: " fmt, chg->name,	\
@@ -1810,6 +1813,13 @@ static int set_sdp_current(struct smb_charger *chg, int icl_ua)
 	u8 icl_options;
 	const struct apsd_result *apsd_result = smblib_get_apsd_result(chg);
 
+#ifdef CONFIG_FORCE_FAST_CHARGE
+	if (force_fast_charge > 0 && icl_ua == USBIN_500MA)
+	{
+		icl_ua = USBIN_900MA;
+	}
+#endif
+
 	/* power source is SDP */
 	switch (icl_ua) {
 	case USBIN_100MA:
diff --git a/drivers/power/supply/qcom_cas/smb1398-charger.c b/drivers/power/supply/qcom_cas/smb1398-charger.c
index 0d3a3e605..acb92b197 100644
--- a/drivers/power/supply/qcom_cas/smb1398-charger.c
+++ b/drivers/power/supply/qcom_cas/smb1398-charger.c
@@ -3764,7 +3764,7 @@ static int smb1398_probe(struct platform_device *pdev)
 	platform_set_drvdata(pdev, chip);
 
 	chip->standalone_mode = true;
-	chip->div2_cp_role = (int)of_device_get_match_data(chip->dev);
+	chip->div2_cp_role = (uintptr_t)of_device_get_match_data(chip->dev);
 	switch (chip->div2_cp_role) {
 	case DIV2_CP_MASTER:
 		rc = smb1398_div2_cp_master_probe(chip);
diff --git a/drivers/power/supply/qcom_cas/smb5-lib.c b/drivers/power/supply/qcom_cas/smb5-lib.c
index c1d5e19e6..011c46882 100644
--- a/drivers/power/supply/qcom_cas/smb5-lib.c
+++ b/drivers/power/supply/qcom_cas/smb5-lib.c
@@ -20,6 +20,9 @@
 #include "step-chg-jeita.h"
 #include "storm-watch.h"
 #include "schgm-flash.h"
+#ifdef CONFIG_FORCE_FAST_CHARGE
+#include <linux/fastchg.h>
+#endif
 
 #define smblib_err(chg, fmt, ...)		\
 	pr_err("%s: %s: " fmt, chg->name,	\
@@ -1759,6 +1762,13 @@ static int set_sdp_current(struct smb_charger *chg, int icl_ua)
 	const struct apsd_result *apsd_result = smblib_get_apsd_result(chg);
 	union power_supply_propval val = {0, };
 
+#ifdef CONFIG_FORCE_FAST_CHARGE
+	if (force_fast_charge > 0 && icl_ua == USBIN_500MA)
+	{
+		icl_ua = USBIN_900MA;
+	}
+#endif
+
 	if (chg->batt_2s_chg) {
 		val.intval = icl_ua;
 		return smblib_set_prop_on_bbc(chg,
diff --git a/drivers/power/supply/ti/bq2597x_charger_dagu.c b/drivers/power/supply/ti/bq2597x_charger_dagu.c
index 3baa1d0db..3b3b2baa6 100644
--- a/drivers/power/supply/ti/bq2597x_charger_dagu.c
+++ b/drivers/power/supply/ti/bq2597x_charger_dagu.c
@@ -1491,9 +1491,9 @@ static int bq2597x_get_dev_role(struct i2c_client *client)
 	}
 
 	dev_info(&client->dev, "%s: matched to %s, dev_role: %d.\n", __func__,
-		 of_id->compatible, (int)of_id->data);
+		 of_id->compatible, (uintptr_t)of_id->data);
 
-	return (int)of_id->data;
+	return (uintptr_t)of_id->data;
 }
 
 static int bq2597x_parse_dt(struct bq2597x *bq, struct device *dev)
diff --git a/drivers/video/backlight/dualktz8866.c b/drivers/video/backlight/dualktz8866.c
index cf034ff66..1387953bc 100644
--- a/drivers/video/backlight/dualktz8866.c
+++ b/drivers/video/backlight/dualktz8866.c
@@ -236,9 +236,6 @@ static int ktz8866_backlight_update_status(struct backlight_device *backlight)
 
     mutex_lock(&g_ktz8866_led.lock);
 
-    dev_warn(&bd->client->dev,
-        "ktz8866 backlight 0x%02x ,exponential brightness %d \n", brightness, exponential_bl);
-
     if (brightness > 0) {
         ktz8866_writes(bd, KTZ8866_DISP_BL_ENABLE, 0x4f);
         dev_warn(&bd->client->dev, "ktz8866 backlight enable,dimming close");
diff --git a/drivers/video/backlight/ktz8866.c b/drivers/video/backlight/ktz8866.c
index 56b3cd880..2deaa8d3e 100755
--- a/drivers/video/backlight/ktz8866.c
+++ b/drivers/video/backlight/ktz8866.c
@@ -75,9 +75,6 @@ static int ktz8866_backlight_update_status(struct backlight_device *backlight)
 
 	mutex_lock(&g_ktz8866_led.lock);
 
-	dev_warn(&bd->client->dev,
-		"ktz8866 backlight 0x%02x ,exponential brightness %d \n", brightness, exponential_bl);
-
 	if (!g_ktz8866_led.ktz8866_status && brightness > 0) {
 		ktz8866_write(KTZ8866_DISP_BL_ENABLE, 0x5f);
 		g_ktz8866_led.ktz8866_status = 1;
diff --git a/include/linux/fastchg.h b/include/linux/fastchg.h
new file mode 100644
index 000000000..b1e4a5d09
--- /dev/null
+++ b/include/linux/fastchg.h
@@ -0,0 +1,22 @@
+/*
+ * Author: Chad Froebel <chadfroebel@gmail.com>
+ *
+ * Port to cheeseburger: engstk <eng.stk@sapo.pt>
+ *
+ * This software is licensed under the terms of the GNU General Public
+ * License version 2, as published by the Free Software Foundation, and
+ * may be copied, distributed, and modified under those terms.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ */
+
+#ifndef _LINUX_FASTCHG_H
+#define _LINUX_FASTCHG_H
+
+extern int force_fast_charge;
+
+#endif
diff --git a/techpack/camera/drivers/cam_sensor_module/cam_ois/cam_ois_core.c b/techpack/camera/drivers/cam_sensor_module/cam_ois/cam_ois_core.c
index 9ab4e2f75..61a03a3fc 100644
--- a/techpack/camera/drivers/cam_sensor_module/cam_ois/cam_ois_core.c
+++ b/techpack/camera/drivers/cam_sensor_module/cam_ois/cam_ois_core.c
@@ -7,6 +7,7 @@
 #include <linux/module.h>
 #include <linux/firmware.h>
 #include <linux/dma-contiguous.h>
+#include <linux/string.h>
 #include <cam_sensor_cmn_header.h>
 #include "cam_ois_core.h"
 #include "cam_ois_soc.h"
@@ -1375,9 +1376,11 @@ static int cam_ois_pkt_parse(struct cam_ois_ctrl_t *o_ctrl, void *arg)
 		if (o_ctrl->ois_fw_flag) {
 			CAM_DBG(CAM_OIS, "is_addr_indata = %d",
 				o_ctrl->opcode.is_addr_indata);
-			if (121 == o_ctrl->opcode.is_addr_indata) {
+			if (121 == o_ctrl->opcode.is_addr_indata ||
+				strnstr(o_ctrl->ois_name, "sem1215", OIS_NAME_LEN)) {
 				CAM_DBG(CAM_OIS,
-					"apply sem1215 ois_fw settings begin.");
+					"apply sem1215 ois_fw settings begin, ois_name: %s",
+					o_ctrl->ois_name);
 				rc = cam_sem1215_ois_fw_download(o_ctrl);
 				CAM_DBG(CAM_OIS,
 					"apply sem1215 ois_fw settings done.");
-- 
2.43.0

